### PR TITLE
Added timestamp calculation to detsim city

### PR DESCRIPTION
**Pull request by MCruces-fz**
Timestamp calculation is based on `rate` parameter and `event number`.

### CODE REVIEW

**Comment by MCruces-fz** [comment]
I have left the modifications of the names of the output files for each detsim test in a different commit at the request of @mmkekic.
All other changes pertinent to the calculation of the timestamp are in the last commit

File: invisible_cities/detsim/sensor_utils.py
Comment (gondiaz) [inline invisible_cities/detsim/sensor_utils.py]: The function output and the warning are inconsistent. Is it 0.5 Hz or 0.5 kHz? Remember that `1 Hz = 1/s`.

File: invisible_cities/detsim/sensor_utils.py
Comment (gondiaz) [inline invisible_cities/detsim/sensor_utils.py]: What if the rate is negative?

File: invisible_cities/detsim/sensor_utils_test.py
Comment (gondiaz) [inline invisible_cities/detsim/sensor_utils_test.py]: Typo (woll -> will). I would rewritte the docstring to something like "Check the rate is always positive/physical".

On the other hand, since you are testing a single value whose output is determined (hardcoded in the function being tested) I think it is better to check the exact result.

File: invisible_cities/cities/components.py
Comment (gondiaz) [inline invisible_cities/cities/components.py:137]: This warning outputs the rate in the default units of IC for 1/time. For example, try to see the warning comment for `rate = -1 * units.hertz`. It does not return the expected warning with positive 1 Hz.

Comment (MCruces-fz) [reply invisible_cities/cities/components.py:137]: Ah, it is true. I should use `abs(rate) / units.hertz`, thanks

**Comment by jacg** [comment]
The tests are failing. It is not immediately obvious to me that the failure is spurious.

+ If the failure is spurious, please rerun the job on IC.
+ If the failure is *not* spurious, please fix: having broken PRs sitting around in the queue is not helpful.

**Comment by jacg** [comment]
... and please rebase on top of #783 as soon as it is merged, so that we get clearer test error reporting.

**Comment by jacg** [comment]
Is there any reason why this PR should not be reviewed and merged?

**Comment by jacg** [comment]
#783 has been merged. Please rebase.

**Comment by MCruces-fz** [comment]
I should clear the commit history a bit. Then, can we merge it?
Maybe I should do more modifications

**Comment by gondiaz** [comment]
This is the warning message I get when running the function in a python shell:
```
>>> f = create_timestamp(-2*units.hertz)
/Users/gonzalo/NEXT/IC/invisible_cities/cities/components.py:137: UserWarning: Negative rate is unphysical, using rate = 2.0 Hz instead
  warnings.warn(f"Negative rate is unphysical, using "
```
Shouldn't we add `stacklevel=2` to the warning? The output would be clearer I think
```
>>> f = create_timestamp(-2*units.hertz)
__main__:1: UserWarning: Negative rate is unphysical, using rate = 2.0 Hz instead
```

**Comment by jacg** [comment]
> I should clear the commit history a bit.

Yes, but there's no point in doing that until you have an approving review ... UNLESS (and it's a *very* important 'unless') cleaning up history will help the reviewers understand what you have done.

Don't hesitate to ask for help with the cleaning, if you need it.

> Then, can we merge it?

I'm not the best person to answer that question. You should probably ask for a review (using the buttons/links on the right, near the top of the PR page) from the people you consider most familiar with the issue you are resolving.

**Comment by MCruces-fz** [comment]
> Shouldn't we add `stacklevel=2` to the warning? [...]

Good idea, I add that now

**Comment by mmkekic** [comment]
> I'm not the best person to answer that question. You should probably ask for a review (using the buttons/links on the right, near the top of the PR page) from the people you consider most familiar with the issue you are resolving.

Well reviewer @gondiaz  has already been assigned and it seems he is mostly happy with the changes.

> Shouldn't we add stacklevel=2 to the warning? [...]

I think this is a good idea, that way the warning will point to the function calling `create_timestamp` where the nonphysical rate value is passed, instead of pointing to the `create_timestamp` itself

**Comment by MCruces-fz** [comment]
Okay, thank you @jacg

**Comment by jacg** [comment]
> Well reviewer @gondiaz has already been assigned and it seems he is mostly happy with the changes.

@gondiaz please formally approve once you are happy, so that it's clear when we can merge.

[@carmenromo You should start thinking about training a successor for the role of merge-master]

**Comment by carmenromo** [comment]
Why? Are you not happy with me? 
I think most of the people know how to merge, but I can teach anyone :)

**Comment by gondiaz** [comment]
The PR will be ready after history rewriting/squashing. The commits should be similar to:

- Modify create_timestamp function
- Move create_timestamp function to components.py
- Add test for create_timestamp function
- Modify detsim city and source with new create_timestamp function
- Modify detsim tests for new create_timestamp function
- Modify buffy city and source with new create_timestamp function

Note that you usually start with a past tense at the start of commit messages, change it to present.

**Comment by jacg** [comment]
> Why? Are you not happy with me?

Very happy!

> I think most of the people know how to merge, but I can teach anyone :)

I'm sure that there are many who can merge, but far fewer are aware of the standards of merge commits we expect in IC.

Also remember that the role of merge-master serves two purposes: the idea is that some junior contributor

1. become comfortable with managing this not-entirely-trivial git operation, and have a regular, not too difficult, but nonetheless important job to do, in order to strengthen a sense of belonging;
2. relieve more senior members of the somewhat mechanical task.

I think that you're waaaaay past needing point 1 for yourself.

If there are any newish team members who would like a little bit of additional responsibility, or an opportunity to get a bit more comfortable with Git, then step forward.

Job description: merge approved PRs as soon as they are ready, following the merge commit style that we have established in IC. Sometimes PRs are merged by senior people involved, but sometimes they rush on to other things. We need someone who will make sure that approved PRs don't stay unmerged for more than 24h (prefereably 24 minutes :-)

**Comment by carmenromo** [comment]
Nice! I see and I fully agree. Do you have someone in mind?

**Comment by jacg** [comment]
> Note that you usually start with a past tense at the start of commit messages, change it to present.

Strictly speaking it's the imperative mood, rather than present tense:

Do this, fix that, remove the other ...

Some people find it helpful to check whether the commit title completes a sentence starting with something like "Including this commit will ..."

For example: Including this commit will ... fix the bug that cost us the Nobel Prize

**Comment by jacg** [comment]
> Do you have someone in mind?

Nope. I have no idea who is lurking in the background, how they feel about their relationship to the project etc. etc.

... maybe, after the pandemic is over and I meet some of the new people ...

**Comment by carmenromo** [comment]
@halmamol has volunteered!

**Comment by jjgomezcadenas** [comment]
What about Miryam?

**Comment by jjgomezcadenas** [comment]
Ja! The usual suspects! @halmamol gets the job!

**Comment by carmenromo** [comment]
If @MiryamMV wants I can teach both and they can agree as I did with @bpalmeiro.

**Comment by jacg** [comment]
Hi @halmamol,

I'm sorry, I don't know who you are. I hope that the coronavirus will allow us to remedy this in the not too distant future. In the meantime don't hesitate to reach out to me if you want any help or assistance, though I am confident that @carmenromo will be able to get you up to speed without any trouble.

**Comment by jacg** [comment]
Hi @MiryamMV,

My last message applies to you too!

**Comment by jacg** [comment]
Sorry @MCruces-fz ... we seem to have hijacked your PR's conversation to discuss the merge-master role.

@carmenromo, @halmamol, @MiryamMV I'll refrain from merging #784 (which is still waiting for @jmbenlloch's approval, but should be ready), to leave you with plenty of material for practice. Please merge that one before any of the others, otherwise the spurious test failures will drive you nuts. (The whole PR is just a single commit, but *do* add a merge commit anyway.)

Next is #780. It's squashed and approved, but wait until #784 is in.

After that you can expect this one (#778) and #781 to be ready soon, so there should be plenty of material for you to work with. Merry merging!

**Comment by MCruces-fz** [comment]
I hadn't realized that my branch was outdated

**Review by gondiaz — APPROVED** [review]
This PR succesfully implements timestamps for MC events in both detsim and buffy cities. Good job!
You can proceed with merging @carmenromo @halmamol @MiryamMV.

**Comment by carmenromo** [comment]
Since the PR #780 is waiting to pass the tests and this PR is already rebased, can we merge this one first @jacg?

**Comment by jacg** [comment]
Yup.

**Pull request merged**
